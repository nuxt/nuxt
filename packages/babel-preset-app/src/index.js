const coreJsMeta = {
  2: {
    prefixes: {
      es6: 'es6',
      es7: 'es7'
    },
    builtIns: require.resolve('@babel/compat-data/corejs2-built-ins')
  },
  3: {
    prefixes: {
      es6: 'es',
      es7: 'es'
    },
    builtIns: require.resolve('core-js-compat/data')
  }
}

function getMajorVersion (version) {
  if (typeof version === 'number') {
    return Math.floor(version)
  } else {
    return Number(version.split('.')[0])
  }
}

function getDefaultPolyfills (corejs) {
  const { prefixes: { es6, es7 } } = coreJsMeta[getMajorVersion(corejs.version)]
  return [
    // Promise polyfill alone doesn't work in IE,
    // Needs this as well. see: #1642
    `${es6}.array.iterator`,
    // This is required for webpack code splitting, vuex etc.
    `${es6}.promise`,
    // this is needed for object rest spread support in templates
    // as vue-template-es2015-compiler 1.8+ compiles it to Object.assign() calls.
    `${es6}.object.assign`,
    // #2012 es7.promise replaces native Promise in FF and causes missing finally
    `${es7}.promise.finally`
  ]
}

function getPolyfills (targets, includes, { ignoreBrowserslistConfig, configPath, corejs }) {
  const { default: getTargets, isRequired } = require('@babel/helper-compilation-targets')
  const builtInsList = require(coreJsMeta[getMajorVersion(corejs.version)].builtIns)
  const builtInTargets = getTargets(targets, {
    ignoreBrowserslistConfig,
    configPath
  })

  return includes.filter(item => isRequired(
    'nuxt-polyfills',
    builtInTargets,
    {
      compatData: { 'nuxt-polyfills': builtInsList[item] }
    }
  ))
}

module.exports = (api, options = {}) => {
  const presets = []
  const plugins = []

  const envName = api.env()

  const {
    bugfixes,
    polyfills: userPolyfills,
    loose = false,
    debug = false,
    useBuiltIns = 'usage',
    modules = false,
    spec,
    ignoreBrowserslistConfig = envName === 'modern',
    configPath,
    include,
    exclude,
    shippedProposals,
    forceAllTransforms,
    decoratorsBeforeExport,
    decoratorsLegacy,
    absoluteRuntime
  } = options

  let { corejs = { version: 2 } } = options

  if (typeof corejs !== 'object') {
    corejs = { version: Number(corejs) }
  }

  if (corejs.proposals === undefined) {
    corejs.proposals = true
  }

  const defaultTargets = {
    server: { node: 'current' },
    client: { ie: 9 },
    modern: { esmodules: true }
  }

  const { targets = defaultTargets[envName] } = options

  const polyfills = []

  if (envName === 'client' && useBuiltIns === 'usage') {
    polyfills.push(
      ...getPolyfills(
        targets,
        userPolyfills || getDefaultPolyfills(corejs),
        {
          ignoreBrowserslistConfig,
          configPath,
          corejs
        }
      )
    )
    plugins.push([require('./polyfills-plugin'), { polyfills }])
  }

  // Pass options along to babel-preset-env
  presets.push([
    require('@babel/preset-env'), {
      bugfixes,
      spec,
      loose,
      debug,
      modules,
      targets,
      useBuiltIns,
      corejs,
      ignoreBrowserslistConfig,
      configPath,
      include,
      exclude: [
        ...exclude || [],
        ...polyfills,
        // Although preset-env includes class-properties
        // but webpack 4 doesn't support the syntax when target supports and babel transpilation is skipped
        // https://github.com/webpack/webpack/issues/9708
        '@babel/plugin-proposal-class-properties',
        '@babel/plugin-proposal-private-methods',
        '@babel/plugin-proposal-private-property-in-object'
      ],
      shippedProposals,
      forceAllTransforms
    }
  ])

  // JSX
  if (options.jsx !== false) {
    presets.push([require('@vue/babel-preset-jsx'), Object.assign({}, options.jsx)])
  }

  plugins.push(
    [require('@babel/plugin-proposal-decorators'), {
      decoratorsBeforeExport,
      legacy: decoratorsLegacy !== false
    }],
    // class-properties and private-methods need same loose value
    [require('@babel/plugin-proposal-class-properties'), { loose: true }],
    [require('@babel/plugin-proposal-private-methods'), { loose: true }],
    [require('@babel/plugin-proposal-private-property-in-object'), { loose: true }]
  )

  // Transform runtime, but only for helpers
  plugins.push([require('@babel/plugin-transform-runtime'), {
    regenerator: useBuiltIns !== 'usage',
    corejs: false,
    helpers: useBuiltIns === 'usage',
    useESModules: envName !== 'server',
    absoluteRuntime
  }])

  // https://github.com/nuxt/nuxt.js/issues/7722
  if (envName === 'server') {
    plugins.push(require('@babel/plugin-proposal-optional-chaining'))
    plugins.push(require('@babel/plugin-proposal-nullish-coalescing-operator'))
  }

  return {
    sourceType: 'unambiguous',
    presets,
    plugins
  }
}
