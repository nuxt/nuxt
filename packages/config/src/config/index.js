import _app from './_app'
import _common from './_common'

import build from './build'
import messages from './messages'
import modes from './modes'
import render from './render'
import router from './router'
import server from './server'
import cli from './cli'
import generate from './generate'

export const defaultNuxtConfigFile = 'nuxt.config'

export function getDefaultNuxtConfig (options = {}) {
  if (!options.env) {
    options.env = process.env
  }

  return {
    ..._app(),
    ..._common(),
    build: build(),
    messages: messages(),
    modes: modes(),
    render: render(),
    router: router(),
    server: server(options),
    cli: cli(),
    generate: generate()
  }
}
