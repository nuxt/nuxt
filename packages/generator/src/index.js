import Generator from './generator'
export { default as Generator } from './generator'

export function getGenerator (nuxt, builder) {
  return new Generator(nuxt, builder)
}
