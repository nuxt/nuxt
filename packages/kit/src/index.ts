// Module
export * from './module/define'
export * from './module/install'

// Loader
export * from './loader/config'
export * from './loader/schema'
export * from './loader/nuxt'

// Utils
export * from './imports'
export * from './build'
export * from './compatibility'
export * from './components'
export * from './context'
export { isIgnored } from './ignore'
export * from './layout'
export * from './pages'
export * from './plugin'
export * from './resolve'
export * from './nitro'
export * from './template'
export * from './logger'

// Internal Utils
// TODO
export {
  resolveModule,
  requireModule,
  importModule,
  tryImportModule,
  tryRequireModule
} from './internal/cjs'
export type { ResolveModuleOptions, RequireModuleOptions } from './internal/cjs'
export { tryResolveModule } from './internal/esm'
export * from './internal/template'
