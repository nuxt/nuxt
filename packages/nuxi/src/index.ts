export * from './run'
export * from './constants'
