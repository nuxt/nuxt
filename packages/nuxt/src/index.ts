export { createNuxt, loadNuxt } from './core/nuxt'
export { build } from './core/builder'
