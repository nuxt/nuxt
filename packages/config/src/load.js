import path from 'path'
import fs from 'fs'
import defu from 'defu'
import consola from 'consola'
import dotenv from 'dotenv'
import { clearRequireCache, createRequire, scanRequireTree } from '@nuxt/utils'
import destr from 'destr'
import * as rc from 'rc9'
import { defaultNuxtConfigFile } from './config'

export async function loadNuxtConfig ({
  rootDir = '.',
  envConfig = {},
  configFile = defaultNuxtConfigFile,
  configContext = {},
  configOverrides = {}
} = {}) {
  rootDir = path.resolve(rootDir)

  const _require = createRequire(rootDir, true)

  let options = {}

  try {
    configFile = _require.resolve(path.resolve(rootDir, configFile))
  } catch (e) {
    if (e.code !== 'MODULE_NOT_FOUND') {
      throw (e)
    } else if (configFile !== defaultNuxtConfigFile) {
      consola.fatal('Config file not found: ' + configFile)
    }
    // Skip configFile if cannot resolve
    configFile = undefined
  }

  // Load env
  envConfig = {
    dotenv: '.env',
    env: process.env,
    expand: true,
    ...envConfig
  }

  const env = loadEnv(envConfig, rootDir)

  // Fill process.env so it is accessible in nuxt.config
  for (const key in env) {
    if (!key.startsWith('_') && envConfig.env[key] === undefined) {
      envConfig.env[key] = env[key]
    }
  }

  if (configFile) {
    // Clear cache
    clearRequireCache(configFile)
    options = _require(configFile) || {}
    if (options.default) {
      options = options.default
    }

    if (typeof options === 'function') {
      try {
        options = await options(configContext)
        if (options.default) {
          options = options.default
        }
      } catch (error) {
        consola.error(error)
        consola.fatal('Error while fetching async configuration')
      }
    }

    // Don't mutate options export
    options = { ...options }

    // Keep _nuxtConfigFile for watching
    options._nuxtConfigFile = configFile

    // Keep all related files for watching
    options._nuxtConfigFiles = Array.from(scanRequireTree(configFile))
    if (!options._nuxtConfigFiles.includes(configFile)) {
      options._nuxtConfigFiles.unshift(configFile)
    }
  }

  if (typeof options.rootDir !== 'string') {
    options.rootDir = rootDir
  }

  // Load Combine configs
  // Priority: configOverrides > nuxtConfig > .nuxt/dist/nuxtrc > .nuxtrc > .nuxtrc (global)
  const dev = configOverrides.dev ?? options.dev ?? configContext.dev
  const buildDir = configOverrides.buildDir ?? options.buildDir ?? configContext.buildDir ?? '.nuxt'
  options = defu(
    configOverrides,
    options,
    !dev ? rc.read({ name: 'nuxtrc', dir: path.resolve(buildDir, 'dist') }) : {},
    rc.read({ name: '.nuxtrc', dir: options.rootDir }),
    rc.readUser('.nuxtrc')
  )

  // Load env to options._env
  options._env = env
  options._envConfig = envConfig
  if (configContext) { configContext.env = env }

  // Expand and interpolate runtimeConfig from _env
  if (envConfig.expand) {
    for (const c of ['publicRuntimeConfig', 'privateRuntimeConfig']) {
      if (options[c]) {
        if (typeof options[c] === 'function') {
          options[c] = options[c](env)
        }
        expand(options[c], env, destr)
      }
    }
  }

  return options
}

function loadEnv (envConfig, rootDir = process.cwd()) {
  const env = Object.create(null)

  // Read dotenv
  if (envConfig.dotenv) {
    envConfig.dotenv = path.resolve(rootDir, envConfig.dotenv)
    if (fs.existsSync(envConfig.dotenv)) {
      const parsed = dotenv.parse(fs.readFileSync(envConfig.dotenv, 'utf-8'))
      Object.assign(env, parsed)
    }
  }

  // Apply process.env
  if (!envConfig.env._applied) {
    Object.assign(env, envConfig.env)
    envConfig.env._applied = true
  }

  // Interpolate env
  if (envConfig.expand) {
    expand(env)
  }

  return env
}

// Based on https://github.com/motdotla/dotenv-expand
function expand (target, source = {}, parse = v => v) {
  function getValue (key) {
    // Source value 'wins' over target value
    return source[key] !== undefined ? source[key] : target[key]
  }

  function interpolate (value, parents = []) {
    if (typeof value !== 'string') {
      return value
    }
    const matches = value.match(/(.?\${?(?:[a-zA-Z0-9_:]+)?}?)/g) || []
    return parse(matches.reduce((newValue, match) => {
      const parts = /(.?)\${?([a-zA-Z0-9_:]+)?}?/g.exec(match)
      const prefix = parts[1]

      let value, replacePart

      if (prefix === '\\') {
        replacePart = parts[0]
        value = replacePart.replace('\\$', '$')
      } else {
        const key = parts[2]
        replacePart = parts[0].substring(prefix.length)

        // Avoid recursion
        if (parents.includes(key)) {
          consola.warn(`Please avoid recursive environment variables ( loop: ${parents.join(' > ')} > ${key} )`)
          return ''
        }

        value = getValue(key)

        // Resolve recursive interpolations
        value = interpolate(value, [...parents, key])
      }

      return value !== undefined ? newValue.replace(replacePart, value) : newValue
    }, value))
  }

  for (const key in target) {
    target[key] = interpolate(getValue(key))
  }
}
