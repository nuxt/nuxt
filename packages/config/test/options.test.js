import path from 'path'
import consola from 'consola'
import { getNuxtConfig } from '../src/options'

jest.mock('std-env', () => ({
  isBrowser: false,
  isTest: true,
  isDevelopment: false,
  isProduction: true,
  isDebug: false,
  isCI: true,
  isWindows: false,
  isMacOS: false,
  isLinux: true
}))

jest.mock('@nuxt/utils', () => ({
  ...jest.requireActual('@nuxt/utils'),
  getMainModule: () => ({ paths: ['/var/nuxt/node_modules'] }),
  getPKG: () => ({ name: 'fake' })
}))

describe('config: options', () => {
  test('should return default nuxt config', () => {
    jest.spyOn(process, 'cwd').mockReturnValue('/var/nuxt/test')
    jest.spyOn(path, 'resolve').mockImplementation((...args) => args.join('/').replace(/\\+/, '/'))
    jest.spyOn(path, 'join').mockImplementation((...args) => args.join('/').replace(/\\+/, '/'))

    const config = getNuxtConfig({
      createRequire: jest.fn(),
      generate: {
        staticAssets: {
          version: 'x'
        }
      }
    })
    expect(config).toMatchSnapshot()

    process.cwd.mockRestore()
    path.resolve.mockRestore()
    path.join.mockRestore()
  })

  test('should prevent duplicate calls with same options', () => {
    const options = {}
    const firstConfig = getNuxtConfig(options)
    const secondConfig = getNuxtConfig(firstConfig)

    expect(firstConfig).toBe(secondConfig)
    expect(firstConfig.__normalized__).toBe(true)
  })

  test('should return default loading when loading is true', () => {
    const { loading } = getNuxtConfig({ loading: true })
    expect(loading).toEqual({
      color: 'black',
      failedColor: 'red',
      height: '2px',
      throttle: 200,
      duration: 5000,
      continuous: false,
      rtl: false,
      css: true
    })
  })

  test('[Compatibility] should transform transition to pageTransition', () => {
    const { pageTransition, transition } = getNuxtConfig({
      transition: 'test-tran'
    })
    expect(pageTransition).toMatchObject({ name: 'test-tran' })
    expect(transition).toBeUndefined()
  })

  test('should transform pageTransition/layoutTransition to name', () => {
    const { pageTransition, layoutTransition } = getNuxtConfig({
      pageTransition: 'test-tran',
      layoutTransition: 'test-layout-tran'
    })
    expect(pageTransition).toMatchObject({ name: 'test-tran' })
    expect(layoutTransition).toMatchObject({ name: 'test-layout-tran' })
  })

  test('should transform extensions to array', () => {
    const { extensions } = getNuxtConfig({ extensions: 'ext' })
    expect(extensions).toEqual(['js', 'mjs', 'ext'])
  })

  test('should support custom global name', () => {
    const { globalName } = getNuxtConfig({ globalName: 'globalNuxt' })
    expect(globalName).toEqual('globalnuxt')
  })

  test('should detect store dir', () => {
    const { store } = getNuxtConfig({ rootDir: path.resolve(__dirname, 'fixtures') })
    expect(store).toEqual(true)
  })

  test('should unset and warn when etag.hash not a function', () => {
    const { render: { etag } } = getNuxtConfig({ render: { etag: { hash: true } } })
    expect(etag).toMatchObject({ hash: undefined })
    expect(consola.warn).not.toHaveBeenCalledWith('render.etag.hash should be a function, received boolean instead')

    const { render: { etag: etagDev } } = getNuxtConfig({ dev: true, render: { etag: { hash: true } } })
    expect(etagDev).toMatchObject({ hash: undefined })
    expect(consola.warn).toHaveBeenCalledWith('render.etag.hash should be a function, received boolean instead')
  })

  test('should enable csp', () => {
    const { render: { csp } } = getNuxtConfig({ render: { csp: { allowedSources: ['/nuxt/*'], test: true } } })
    expect(csp).toEqual({
      hashAlgorithm: 'sha256',
      addMeta: false,
      unsafeInlineCompatibility: false,
      allowedSources: ['/nuxt/*'],
      policies: undefined,
      reportOnly: false,
      test: true,
      generateNonce: false
    })
  })

  // TODO: Remove this test in Nuxt 3, we will stop supporting this typo (more on: https://github.com/nuxt/nuxt.js/pull/6583)
  test('should enable csp with old typo property name, avoiding breaking changes', () => {
    const { render: { csp } } = getNuxtConfig({ render: { csp: { allowedSources: ['/nuxt/*'], test: true, unsafeInlineCompatiblity: true } } })
    expect(csp).toEqual({
      hashAlgorithm: 'sha256',
      addMeta: false,
      unsafeInlineCompatibility: true,
      allowedSources: ['/nuxt/*'],
      policies: undefined,
      reportOnly: false,
      test: true,
      generateNonce: false
    })
  })

  test('should fallback to server target', () => {
    const { target } = getNuxtConfig({ target: 0 })
    expect(target).toEqual('server')
  })

  test('should check unknown target', () => {
    const { target } = getNuxtConfig({ target: 'test' })
    expect(consola.warn).toHaveBeenCalledWith('Unknown target: test. Falling back to server')
    expect(target).toEqual('server')
  })

  test('should check unknown mode', () => {
    const { build, render } = getNuxtConfig({ mode: 'test' })
    expect(consola.warn).toHaveBeenCalledWith('Unknown mode: test. Falling back to universal')
    expect(build.ssr).toEqual(true)
    expect(render.ssr).toEqual(true)
  })

  test('should add appear true in pageTransition when no ssr', () => {
    const { pageTransition } = getNuxtConfig({ render: { ssr: false } })
    expect(pageTransition.appear).toEqual(true)
  })

  test('should return 200.html as default generate.fallback', () => {
    const { generate: { fallback } } = getNuxtConfig({})
    expect(fallback).toEqual('200.html')
  })

  test('should return 404.html when generate.fallback is true', () => {
    const { generate: { fallback } } = getNuxtConfig({ generate: { fallback: true } })
    expect(fallback).toEqual('404.html')
  })

  test('should return fallback html when generate.fallback is string', () => {
    const { generate: { fallback } } = getNuxtConfig({ generate: { fallback: 'fallback.html' } })
    expect(fallback).toEqual('fallback.html')
  })

  test('export should alias to generate', () => {
    const { generate: { fallback } } = getNuxtConfig({ export: { fallback: 'fallback.html' } })
    expect(fallback).toEqual('fallback.html')
  })

  test('should disable parallel if extractCSS is enabled', () => {
    const { build: { parallel } } = getNuxtConfig({ build: { extractCSS: true, parallel: true } })
    expect(parallel).toEqual(false)
    expect(consola.warn).toHaveBeenCalledWith('extractCSS cannot work with parallel build due to limited work pool in thread-loader')
  })

  describe('config: router dir', () => {
    test('should transform middleware to array', () => {
      const { router: { middleware } } = getNuxtConfig({ router: { middleware: 'midd' } })
      expect(middleware).toEqual(['midd'])
    })

    test('should set _routerBaseSpecified when base is specified', () => {
      const { _routerBaseSpecified } = getNuxtConfig({ router: { base: '/test' } })
      expect(_routerBaseSpecified).toEqual(true)
    })
  })

  describe('config: options dir', () => {
    test('should support custom root dir', () => {
      const { rootDir } = getNuxtConfig({
        rootDir: 'root'
      })
      expect(rootDir).toEqual(path.resolve('root'))
    })

    test('should support custom src dir', () => {
      const { srcDir } = getNuxtConfig({
        rootDir: 'root',
        srcDir: 'src'
      })
      expect(srcDir).toEqual(path.resolve('root', 'src'))
    })

    test('should support custom generate dir', () => {
      const { generate: { dir } } = getNuxtConfig({
        rootDir: 'root',
        generate: { dir: 'generate' }
      })
      expect(dir).toEqual(path.resolve('root', 'generate'))
    })
  })

  describe('config: options template', () => {
    test('should use default appTemplatePath', () => {
      const { appTemplatePath } = getNuxtConfig({})
      expect(appTemplatePath).toEqual(path.resolve('.nuxt', 'views', 'app.template.html'))
    })

    test('should use custom appTemplatePath', () => {
      const { appTemplatePath } = getNuxtConfig({ appTemplatePath: 'templates' })
      expect(appTemplatePath).toEqual(path.resolve('templates'))
    })

    test('should use custom app.html', () => {
      const { appTemplatePath } = getNuxtConfig({ rootDir: path.resolve(__dirname, 'fixtures') })
      expect(appTemplatePath).toEqual(path.resolve(__dirname, 'fixtures', 'app.html'))
    })
  })

  describe('config: options publicPath', () => {
    test('should fallback to default when publicPath is falsy', () => {
      const { build: { publicPath } } = getNuxtConfig({ build: { publicPath: false } })
      expect(publicPath).toEqual('/_nuxt/')
    })

    test('should append slash in publicPath', () => {
      const { build: { publicPath } } = getNuxtConfig({ build: { publicPath: '/nuxt_public' } })
      expect(publicPath).toEqual('/nuxt_public/')
    })

    test('should ignore url publicPath in dev', () => {
      const { build: { publicPath } } = getNuxtConfig({ dev: true, build: { publicPath: 'http://nuxt_public' } })
      expect(publicPath).toEqual('/_nuxt/')
    })
  })

  describe('config: options babel', () => {
    test('should replace and deprecate @nuxtjs/babel-preset-app', () => {
      const { build: { babel } } = getNuxtConfig({
        build: { babel: { presets: ['@nuxtjs/babel-preset-app'] } }
      })
      expect(consola.warn).toHaveBeenCalledWith('@nuxtjs/babel-preset-app has been deprecated, please use @nuxt/babel-preset-app.')
      expect(babel).toEqual({
        configFile: false,
        babelrc: false,
        cacheDirectory: false,
        presets: ['@nuxt/babel-preset-app']
      })
    })

    test('should support options in babel presets', () => {
      const { build: { babel } } = getNuxtConfig({
        build: { babel: { presets: [['@nuxt/babel-preset-app', { test: true }]] } }
      })
      expect(babel).toEqual({
        configFile: false,
        babelrc: false,
        cacheDirectory: false,
        presets: [['@nuxt/babel-preset-app', { test: true }]]
      })
    })
  })

  describe('config: options deprecated', () => {
    test('should deprecate render.gzip', () => {
      getNuxtConfig({ render: { gzip: true } })
      expect(consola.warn).toHaveBeenCalledWith('render.gzip is deprecated and will be removed in a future version! Please switch to render.compressor')
    })

    test('should deprecate build.vendor', () => {
      getNuxtConfig({ build: { vendor: ['lodash'] } })
      expect(consola.warn).toHaveBeenCalledWith('vendor has been deprecated due to webpack4 optimization')
    })

    test('should deprecate devModules', () => {
      const config = getNuxtConfig({ devModules: ['foo'], buildModules: ['bar'] })
      expect(consola.warn).toHaveBeenCalledWith('`devModules` has been renamed to `buildModules` and will be removed in Nuxt 3.')
      expect(config.devModules).toBe(undefined)
      expect(config.buildModules).toEqual(['bar', 'foo'])
    })

    test('should deprecate build.extractCSS.allChunks', () => {
      getNuxtConfig({ build: { extractCSS: { allChunks: true } } })
      expect(consola.warn).toHaveBeenCalledWith('build.extractCSS.allChunks has no effect from v2.0.0. Please use build.optimization.splitChunks settings instead.')
    })

    test('should deprecate build.crossorigin', () => {
      getNuxtConfig({ build: { crossorigin: 'use-credentials' } })
      expect(consola.warn).toHaveBeenCalledWith('Using `build.crossorigin` is deprecated and will be removed in Nuxt 3. Please use `render.crossorigin` instead.')
    })
  })
})

describe('config: serverMiddleware', () => {
  test('should transform serverMiddleware hash', () => {
    const serverMiddleware = {
      '/resource': (req, res, next) => {
      }
    }
    const config = getNuxtConfig({ serverMiddleware })
    expect(config.serverMiddleware[0].path).toBe('/resource')
    expect(config.serverMiddleware[0].handler).toBe(serverMiddleware['/resource'])
  })
})

describe('config: router', () => {
  test('should sanitize router.base', () => {
    const config = getNuxtConfig({ router: { base: '/foo' } })
    expect(config.router.base).toBe('/foo/')
  })
})
