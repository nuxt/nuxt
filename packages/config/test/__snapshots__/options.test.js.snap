// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`config: options should return default nuxt config 1`] = `
Object {
  "ErrorPage": null,
  "__normalized__": true,
  "_modules": Array [
    "@nuxt/components",
  ],
  "_nuxtConfigFile": "/var/nuxt/test/nuxt.config.js",
  "_nuxtConfigFiles": Array [
    "/var/nuxt/test/nuxt.config.js",
  ],
  "alias": Object {
    "@": "/var/nuxt/test",
    "@@": "/var/nuxt/test",
    "assets": "/var/nuxt/test/assets",
    "static": "/var/nuxt/test/static",
    "~": "/var/nuxt/test",
    "~~": "/var/nuxt/test",
  },
  "app": Object {
    "assetsPath": "/_nuxt/",
    "basePath": "/",
    "cdnURL": null,
  },
  "appTemplatePath": "/var/nuxt/test/.nuxt/views/app.template.html",
  "build": Object {
    "_publicPath": "/_nuxt/",
    "additionalExtensions": Array [],
    "aggressiveCodeRemoval": false,
    "analyze": false,
    "babel": Object {
      "babelrc": false,
      "cacheDirectory": false,
      "configFile": false,
    },
    "cache": false,
    "corejs": "auto",
    "cssSourceMap": false,
    "devMiddleware": Object {
      "stats": "none",
    },
    "extractCSS": false,
    "filenames": Object {
      "app": [Function],
      "chunk": [Function],
      "css": [Function],
      "font": [Function],
      "img": [Function],
      "video": [Function],
    },
    "followSymlinks": false,
    "friendlyErrors": true,
    "hardSource": false,
    "hotMiddleware": Object {},
    "html": Object {
      "minify": Object {
        "collapseBooleanAttributes": true,
        "decodeEntities": true,
        "minifyCSS": true,
        "minifyJS": true,
        "processConditionalComments": true,
        "removeEmptyAttributes": true,
        "removeRedundantAttributes": true,
        "trimCustomFragments": true,
        "useShortDoctype": true,
      },
    },
    "indicator": false,
    "loaders": Object {
      "css": Object {
        "esModule": false,
        "modules": Object {
          "compileType": "icss",
        },
        "sourceMap": false,
      },
      "cssModules": Object {
        "esModule": false,
        "modules": Object {
          "localIdentName": "[local]_[hash:base64:5]",
        },
        "sourceMap": false,
      },
      "file": Object {
        "esModule": false,
      },
      "fontUrl": Object {
        "esModule": false,
        "limit": 1000,
      },
      "imgUrl": Object {
        "esModule": false,
        "limit": 1000,
      },
      "less": Object {
        "sourceMap": false,
      },
      "pugPlain": Object {},
      "sass": Object {
        "sassOptions": Object {
          "indentedSyntax": true,
        },
        "sourceMap": false,
      },
      "scss": Object {
        "sourceMap": false,
      },
      "stylus": Object {
        "sourceMap": false,
      },
      "vue": Object {
        "productionMode": true,
        "transformAssetUrls": Object {
          "embed": "src",
          "object": "src",
          "source": "src",
          "video": "src",
        },
      },
      "vueStyle": Object {
        "sourceMap": false,
      },
    },
    "loadingScreen": false,
    "optimization": Object {
      "minimize": true,
      "minimizer": undefined,
      "runtimeChunk": "single",
      "splitChunks": Object {
        "automaticNameDelimiter": "/",
        "cacheGroups": Object {},
        "chunks": "all",
      },
    },
    "optimizeCSS": false,
    "parallel": false,
    "plugins": Array [],
    "postcss": Object {
      "preset": Object {
        "stage": 2,
      },
    },
    "profile": false,
    "publicPath": "/_nuxt/",
    "quiet": true,
    "serverURLPolyfill": "url",
    "splitChunks": Object {
      "commons": true,
      "layouts": false,
      "pages": true,
    },
    "ssr": true,
    "standalone": false,
    "stats": false,
    "styleResources": Object {},
    "template": undefined,
    "templates": Array [],
    "terser": Object {},
    "transpile": Array [],
    "warningIgnoreFilters": Array [],
    "watch": Array [],
  },
  "buildDir": "/var/nuxt/test/.nuxt",
  "buildModules": Array [],
  "cli": Object {
    "badgeMessages": Array [],
    "bannerColor": "green",
  },
  "createRequire": [MockFunction],
  "css": Array [],
  "debug": false,
  "dev": false,
  "dir": Object {
    "app": "app",
    "assets": "assets",
    "layouts": "layouts",
    "middleware": "middleware",
    "pages": "pages",
    "static": "static",
    "store": "store",
  },
  "editor": undefined,
  "env": Object {},
  "extendPlugins": null,
  "extensions": Array [
    "js",
    "mjs",
  ],
  "features": Object {
    "asyncData": true,
    "clientOnline": true,
    "clientPrefetch": true,
    "componentAliases": true,
    "componentClientOnly": true,
    "deprecations": true,
    "fetch": true,
    "layouts": true,
    "meta": true,
    "middleware": true,
    "store": true,
    "transitions": true,
    "validate": true,
  },
  "fetch": Object {
    "client": true,
    "server": true,
  },
  "generate": Object {
    "cache": Object {
      "globbyOptions": Object {
        "gitignore": true,
      },
      "ignore": Array [],
    },
    "concurrency": 500,
    "crawler": true,
    "dir": "/var/nuxt/test/dist",
    "exclude": Array [],
    "fallback": "200.html",
    "ignoreEnv": false,
    "interval": 0,
    "manifest": true,
    "nojekyll": true,
    "routes": Array [],
    "staticAssets": Object {
      "base": "/_nuxt/static",
      "dir": "static",
      "version": "x",
      "versionBase": "/_nuxt/static/x",
    },
    "subFolders": true,
  },
  "globalName": "nuxt",
  "globals": Object {
    "context": [Function],
    "id": [Function],
    "loadedCallback": [Function],
    "nuxt": [Function],
    "pluginPrefix": [Function],
    "readyCallback": [Function],
  },
  "head": Object {
    "link": Array [],
    "meta": Array [],
    "script": Array [],
    "style": Array [],
  },
  "hooks": null,
  "ignore": Array [
    "**/*.test.*",
    "**/*.spec.*",
    "**/-*.*",
  ],
  "ignoreOptions": undefined,
  "ignorePrefix": "-",
  "layoutTransition": Object {
    "mode": "out-in",
    "name": "layout",
  },
  "layouts": Object {},
  "loading": Object {
    "color": "black",
    "continuous": false,
    "css": true,
    "duration": 5000,
    "failedColor": "red",
    "height": "2px",
    "rtl": false,
    "throttle": 200,
  },
  "loadingIndicator": Object {
    "background": "white",
    "color": "black",
    "color2": "#F5F5F5",
    "dev": false,
    "loading": "Loading...",
    "name": "default",
  },
  "messages": Object {
    "back_to_home": "Back to the home page",
    "client_error": "Error",
    "client_error_details": "An error occurred while rendering the page. Check developer tools console for details.",
    "error_404": "This page could not be found",
    "loading": "Loading...",
    "nuxtjs": "Nuxt",
    "server_error": "Server error",
    "server_error_details": "An error occurred in the application and your page could not be served. If you are the application owner, check your logs for details.",
  },
  "mode": "universal",
  "modern": undefined,
  "modes": Object {
    "spa": Object {
      "build": Object {
        "ssr": false,
      },
      "render": Object {
        "ssr": false,
      },
    },
    "universal": Object {
      "build": Object {
        "ssr": true,
      },
      "render": Object {
        "ssr": true,
      },
    },
  },
  "modules": Array [],
  "modulesDir": Array [
    "/var/nuxt/node_modules",
    "/var/nuxt/test/node_modules",
  ],
  "pageTransition": Object {
    "appear": false,
    "appearActiveClass": "appear-active",
    "appearClass": "appear",
    "appearToClass": "appear-to",
    "mode": "out-in",
    "name": "page",
  },
  "plugins": Array [],
  "privateRuntimeConfig": Object {},
  "publicRuntimeConfig": Object {
    "_app": Object {
      "assetsPath": "/_nuxt/",
      "basePath": "/",
      "cdnURL": null,
    },
  },
  "render": Object {
    "bundleRenderer": Object {
      "runInNewContext": false,
      "shouldPrefetch": [Function],
      "shouldPreload": [Function],
    },
    "compressor": Object {
      "threshold": 0,
    },
    "crossorigin": undefined,
    "csp": false,
    "dist": Object {
      "index": false,
      "maxAge": "1y",
    },
    "etag": Object {
      "weak": false,
    },
    "fallback": Object {
      "dist": Object {},
      "static": Object {
        "handlers": Object {
          ".htm": false,
          ".html": false,
        },
        "skipUnknown": true,
      },
    },
    "http2": Object {
      "push": false,
      "pushAssets": null,
      "shouldPush": null,
    },
    "resourceHints": true,
    "ssr": true,
    "ssrLog": false,
    "static": Object {
      "prefix": true,
    },
  },
  "rootDir": "/var/nuxt/test",
  "router": Object {
    "base": "/",
    "extendRoutes": null,
    "fallback": false,
    "linkActiveClass": "nuxt-link-active",
    "linkExactActiveClass": "nuxt-link-exact-active",
    "linkPrefetchedClass": false,
    "middleware": Array [],
    "mode": "history",
    "parseQuery": false,
    "prefetchLinks": true,
    "prefetchPayloads": true,
    "routeNameSplitter": "-",
    "routes": Array [],
    "scrollBehavior": null,
    "stringifyQuery": false,
    "trailingSlash": undefined,
  },
  "server": Object {
    "host": "localhost",
    "https": false,
    "port": 3000,
    "socket": undefined,
    "timing": false,
  },
  "serverMiddleware": Array [],
  "srcDir": "/var/nuxt/test",
  "ssr": true,
  "styleExtensions": Array [
    "css",
    "pcss",
    "postcss",
    "styl",
    "stylus",
    "scss",
    "sass",
    "less",
  ],
  "target": "server",
  "test": true,
  "vue": Object {
    "config": Object {
      "performance": false,
      "silent": true,
    },
  },
  "vueMeta": null,
  "watch": Array [
    "/var/nuxt/test/nuxt.config.js",
  ],
  "watchers": Object {
    "chokidar": Object {
      "ignoreInitial": true,
    },
    "rewatchOnRawEvents": undefined,
    "webpack": Object {
      "aggregateTimeout": 1000,
    },
  },
}
`;
